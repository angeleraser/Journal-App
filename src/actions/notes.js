import Swal from "sweetalert2";
import { dataBase } from "../firebase/firebase-config";
import { uploadFile } from "../helpers/fileUpload";
import { fetchUserInfoFromDatabase } from "../helpers/loadNotes";
import { TYPES } from "../types/types";

export const addNewNote = () => {
  return async (dispatch, getState) => {
    // getState returns the current main state and it can be used in async operations
    const {
      auth: { uID },
    } = getState();
    const newEntry = {
      body: "",
      title: "",
      date: new Date().getTime(),
      imgURL: null,
    };
    /*
     Collection journal by user ID 
    */
    const DOCUMENT_REF = await dataBase
      .collection(`${uID}/journal/notes`)
      .add(newEntry);
    dispatch(selectNote(DOCUMENT_REF.id, newEntry));
    dispatch(updateNoteList(DOCUMENT_REF.id, newEntry));
  };
};

export const selectNote = (id, note) => {
  return {
    type: TYPES.notesSelect,
    payload: {
      id,
      ...note,
    },
  };
};

export const setNotes = (notes) => {
  return {
    type: TYPES.notesLoad,
    payload: notes,
  };
};

export const startLoadingNotes = (uID) => {
  return async (dispatch) => {
    const notes = await fetchUserInfoFromDatabase(uID, "journal/notes");
    dispatch(setNotes(notes));
  };
};

export const startSaveNote = (note) => {
  return async (dispatch, getState) => {
    const {
      auth: { uID },
    } = getState();
    const noteToFirestore = { ...note };
    delete noteToFirestore.id;
    await dataBase
      .doc(`${uID}/journal/notes/${note.id}`)
      .update(noteToFirestore);
    dispatch(refreshNote(note.id, noteToFirestore));
    Swal.fire("Saved", note.title, "success");
  };
};

export const refreshNote = (id, note) => ({
  type: TYPES.notesUpdate,
  payload: {
    id,
    note,
  },
});

export const startUploadingImage = (file) => {
  return async (dispatch, getState) => {
    const { selected } = getState().notes;
    Swal.fire({
      title: "Uploading...",
      text: "Please wait a few seconds...",
      allowOutsideClick: false,
      onBeforeOpen() {
        Swal.showLoading();
      },
    });
    const fileURL = await uploadFile(file);
    selected.imgURL = fileURL;
    Swal.close();
    dispatch(startSaveNote(selected));
  };
};

export const startDeletingNote = (id) => {
  return async (dispatch, getState) => {
    const uID = getState().auth.uID;
    dataBase.doc(`${uID}/journal/notes/${id}`).delete();
    dispatch(deleteNote(id));
  };
};

export const deleteNote = (id) => ({
  type: TYPES.notesDelete,
  payload: {
    id,
  },
});

export const cleanNotesAfterLogout = () => ({
  type: TYPES.notesLogoutCleaning,
});

export const updateNoteList = (id, note) => {
  return {
    type: TYPES.notesAddNew,
    payload: {
      id,
      ...note,
    },
  };
};
