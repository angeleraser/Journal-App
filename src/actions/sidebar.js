import { TYPES } from "../types/types";

export const openSidebar = () => ({
  type: TYPES.sidebarOpen,
});

export const closedSidebar = () => ({
  type: TYPES.sidebarClosed,
});
