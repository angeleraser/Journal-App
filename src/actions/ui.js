import { TYPES } from "../types/types";

export const setUiError = (errorMsg) => {
  return {
    type: TYPES.uiSetError,
    payload: errorMsg,
  };
};

export const removeUiError = () => {
  return {
    type: TYPES.uiRemoveError,
  };
};
