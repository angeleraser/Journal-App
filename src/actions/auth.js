import Swal from "sweetalert2";
import { FIREBASE, googleAuthProvider } from "..";
import { TYPES } from "../types/types";
import { uiStartLoading, uiFinishLoading } from "./ui";
// NOTA: Si la tarea es asincrona la funcion debe retonar otra funcion que realiza el dispatch, sino un objeto con el tipo de accion y el payload normal.

export const login = (uID, displayName) => ({
  type: TYPES.login,
  payload: {
    uID,
    displayName,
  },
});

export const startLoginEmailPassword = (email, password) => {
  return async (dispatch) => {
    try {
      dispatch(uiStartLoading());
      {
        const { user } = await FIREBASE.auth().signInWithEmailAndPassword(
          email,
          password
        );
        dispatch(login(user.uid, user.displayName));
      }
    } catch (error) {
      Swal.fire("Error", error.message, "error");
      console.log(error);
    } finally {
      dispatch(uiFinishLoading());
    }
  };
};

export const startRegisterWithEmailPasswordName = (email, password, name) => {
  return async (dispatch) => {
    try {
      dispatch(uiStartLoading());
      {
        const { user } = await FIREBASE.auth().createUserWithEmailAndPassword(
          email,
          password
        );
        await user.updateProfile({ displayName: name });
        dispatch(login(user.uid, user.displayName));
      }
    } catch (error) {
      console.log(error);
      Swal.fire("Error", error.message, "error");
    } finally {
      dispatch(uiFinishLoading());
    }
  };
};

export const startGoogleLogin = () => {
  return async (dispatch) => {
    try {
      dispatch(uiStartLoading());
      {
        const { user } = await FIREBASE.auth().signInWithPopup(
          googleAuthProvider
        );
        dispatch(login(user.id, user.displayName));
      }
    } catch (error) {
      console.log(error);
      Swal.fire("Error", error.message, "error");
    } finally {
      dispatch(uiFinishLoading());
    }
  };
};

export const startLogout = () => {
  return async (dispatch) => {
    await FIREBASE.auth().signOut();
    dispatch(logout());
  };
};

export const logout = () => ({
  type: TYPES.logout,
});
