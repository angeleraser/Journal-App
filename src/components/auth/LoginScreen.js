import React from "react";
import { Link } from "react-router-dom";
import { googleIcon } from "../../icons/svg";
import useForm from "../hooks/useForm";
import { useDispatch, useSelector } from "react-redux";
import { startGoogleLogin, startLoginEmailPassword } from "../../actions/auth";
export const LoginScreen = () => {
  const [{ email, password }, handleInputChange] = useForm({
    email: "",
    password: "",
  });
  const dispatch = useDispatch();
  const { loading } = useSelector(({ ui }) => ui);
  const handleLogin = (e) => {
    e.preventDefault();
    dispatch(startLoginEmailPassword(email, password));
  };
  const handleGoogleLogin = () => {
    dispatch(startGoogleLogin());
  };
  return (
    <>
      <h3 className="auth__title">Login</h3>
      <form
        onSubmit={handleLogin}
        autoComplete="off"
        className="flex flex-col w-full items-start">
        <input
          className="auth__input"
          type="email"
          name="email"
          placeholder="email@example.com"
          value={email}
          onChange={handleInputChange}
        />
        <input
          type="password"
          name="password"
          placeholder="Enter your password..."
          className="auth__input"
          value={password}
          onChange={handleInputChange}
        />
        <button
          disabled={loading}
          className="btn btn-primary btn-block"
          type="submit">
          {!loading ? "Login" : "Loading..."}
        </button>
        <div className="auth__social-networks">
          <div className="networks-buttons w-full mb-4">
            <button
              type="button"
              onClick={handleGoogleLogin}
              className="google-btn-wrapper w-full flex items-center">
              <div className="google-icon mr-10"> {googleIcon}</div>
              Sign in with Google
            </button>
          </div>
        </div>
        <Link className="link font-bold w-full text-center" to="/auth/register">
          Create new account
        </Link>
      </form>
    </>
  );
};
