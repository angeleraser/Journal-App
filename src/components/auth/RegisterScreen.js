import React from "react";
import { Link } from "react-router-dom";
import useForm from "../hooks/useForm";
import { useDispatch, useSelector } from "react-redux";
import { isFormValid } from "../../helpers/isFormValid.js";
import { startRegisterWithEmailPasswordName } from "../../actions/auth";
export const RegisterScreen = () => {
  const [
    { name, email, password, confirmPassword },
    handleInputChange,
  ] = useForm({
    name: "",
    email: "",
    password: "",
    confirmPassword: "",
  });
  const dispatch = useDispatch();
  const { errorMsg } = useSelector((state) => state.ui);
  const handleRegister = (e) => {
    e.preventDefault();
    if (
      isFormValid(
        { name, email, password, confirmPassword },
        dispatch,
        errorMsg
      )
    ) {
      dispatch(startRegisterWithEmailPasswordName(email, password, name));
    }
  };
  return (
    <>
      {" "}
      <h3 className="auth__title">Register</h3>
      {!!errorMsg && (
        <p className="auth__invalid-form-msg w-full text-center font-bold mb-4">
          {" "}
          {errorMsg}
        </p>
      )}
      <form
        onSubmit={handleRegister}
        autoComplete="off"
        className="flex flex-col w-full items-start">
        <input
          className="auth__input"
          type="name"
          value={name}
          name="name"
          onChange={handleInputChange}
          placeholder="Your name"
        />
        <input
          className="auth__input"
          value={email}
          type="email"
          name="email"
          onChange={handleInputChange}
          placeholder="email@example.com"
        />
        <input
          type="password"
          name="password"
          value={password}
          onChange={handleInputChange}
          placeholder="Enter your password..."
          className="auth__input"
        />
        <input
          type="password"
          value={confirmPassword}
          name="confirmPassword"
          onChange={handleInputChange}
          placeholder="Confirm your password..."
          className="auth__input"
        />
        <button className="btn mb-6 btn-primary btn-block" type="submit">
          Register
        </button>
        <Link className="link" to="/auth/login">
          Already register?
        </Link>
      </form>
    </>
  );
};
