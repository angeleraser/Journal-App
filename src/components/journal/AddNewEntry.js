import React from "react";
import { useDispatch } from "react-redux";
import { addNewNote } from "../../actions/notes";
import { closedSidebar } from "../../actions/sidebar";
import { addIconSvg } from "../../icons/svg";

export const AddNewEntry = () => {
  const dispatch = useDispatch();
  return (
    <button
      onClick={() => {
        dispatch(addNewNote());
        dispatch(closedSidebar());
      }}
      className="journal__new-entry">
      Add New entry
      {addIconSvg}
    </button>
  );
};
