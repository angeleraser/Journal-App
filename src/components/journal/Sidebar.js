import React from "react";
import { useSelector } from "react-redux";
import { AddNewEntry } from "./AddNewEntry";
import { JournalEntries } from "./JournalEntries";
import { JournalNavbar } from "./JournalNavbar";

export const Sidebar = () => {
  const { isOpen } = useSelector(({ sidebar }) => sidebar);
  return (
    <aside className={`journal__sidebar ${isOpen ? "active" : ""}`}>
      <JournalNavbar />
      <AddNewEntry />
      <JournalEntries />
    </aside>
  );
};
