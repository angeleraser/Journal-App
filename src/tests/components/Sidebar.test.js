import React from "react";
import { mount } from "enzyme";
import configureStore from "redux-mock-store";
import thunk from "redux-thunk";
import { Provider } from "react-redux";
import { MemoryRouter } from "react-router-dom";
import "@testing-library/jest-dom";
import { Sidebar } from "../../components/journal/Sidebar";
import { startLogout } from "../../actions/auth";
import { addNewNote } from "../../actions/notes";
const middlewares = [thunk];
const mockStore = configureStore(middlewares);
// Hay que definir el estado inicial igual al de la aplicacion en ese punto
const initState = {
  auth: {
    name: "angeleraser",
    uID: "k1j2g5bk75j23",
    photoURL: null,
  },
  notes: {
    notes: [],
    selected: null,
  },
  ui: {
    loading: false,
    errorMsg: null,
  },
  sidebar: {
    isOpen: false,
  },
};
let store = mockStore(initState);
store.dispatch = jest.fn()
const wrapper = mount(
  <Provider store={store}>
    <MemoryRouter>
      <Sidebar />
    </MemoryRouter>
  </Provider>
);
// MOCK AUTH
jest.mock("../../actions/auth", () => {
  return {
    startLogout: jest.fn(),
  };
});
// MOCK NOTES
jest.mock("../../actions/notes", () => {
  return {
    addNewNote: jest.fn(),
  };
});
describe("Testing <Sidebar/>", () => {
  // 1-
  test("should match with snap", () => {
    expect(wrapper).toMatchSnapshot();
  });
  //  2 -
  test("should call startLogout action", () => {
    const logoutButton = wrapper.find(".logout-button");
    // startLogout.mockReturnValue({ type: TYPES.logout });
    logoutButton.prop("onClick")();
    expect(logoutButton.exists()).toBe(true);
    expect(startLogout).toHaveBeenCalled();
  });
  // 3 -
  test("should call startNewNote action", () => {
    const addEntryButton = wrapper.find(".journal__new-entry");
    // addNewNote.mockReturnValue({ type: TYPES.notesAddNew });
    addEntryButton.prop("onClick")();
    expect(addEntryButton.exists()).toBe(true);
    expect(addNewNote).toHaveBeenCalled();
  });
});
