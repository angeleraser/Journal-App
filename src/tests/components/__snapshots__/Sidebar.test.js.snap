// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing <Sidebar/> should match with snap 1`] = `
<aside
  className="journal__sidebar "
>
  <div
    className="journal__sidebar-navbar relative"
  >
    <button
      className="menu-btn"
      onClick={[Function]}
    >
      <svg
        height="24"
        version="1.1"
        viewBox="0 0 32 32"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M31.708 25.708c-0-0-0-0-0-0l-9.708-9.708 9.708-9.708c0-0 0-0 0-0 0.105-0.105 0.18-0.227 0.229-0.357 0.133-0.356 0.057-0.771-0.229-1.057l-4.586-4.586c-0.286-0.286-0.702-0.361-1.057-0.229-0.13 0.048-0.252 0.124-0.357 0.228 0 0-0 0-0 0l-9.708 9.708-9.708-9.708c-0-0-0-0-0-0-0.105-0.104-0.227-0.18-0.357-0.228-0.356-0.133-0.771-0.057-1.057 0.229l-4.586 4.586c-0.286 0.286-0.361 0.702-0.229 1.057 0.049 0.13 0.124 0.252 0.229 0.357 0 0 0 0 0 0l9.708 9.708-9.708 9.708c-0 0-0 0-0 0-0.104 0.105-0.18 0.227-0.229 0.357-0.133 0.355-0.057 0.771 0.229 1.057l4.586 4.586c0.286 0.286 0.702 0.361 1.057 0.229 0.13-0.049 0.252-0.124 0.357-0.229 0-0 0-0 0-0l9.708-9.708 9.708 9.708c0 0 0 0 0 0 0.105 0.105 0.227 0.18 0.357 0.229 0.356 0.133 0.771 0.057 1.057-0.229l4.586-4.586c0.286-0.286 0.362-0.702 0.229-1.057-0.049-0.13-0.124-0.252-0.229-0.357z"
        />
      </svg>
    </button>
    <div
      className="flex items-center user-profile"
    >
      <span>
        angeleraser
      </span>
    </div>
    <button
      className="btn logout-button"
      onClick={[Function]}
    >
      Logout
    </button>
  </div>
  <button
    className="journal__new-entry"
    onClick={[Function]}
  >
    Add New entry
    <svg
      id="Capa_1"
      style={
        Object {
          "enableBackground": "new 0 0 42 42",
        }
      }
      version="1.1"
      viewBox="0 0 42 42"
      x="0px"
      xmlSpace="preserve"
      xmlns="http://www.w3.org/2000/svg"
      xmlnsXlink="http://www.w3.org/1999/xlink"
      y="0px"
    >
      <path
        d="M37.059,16H26V4.941C26,2.224,23.718,0,21,0s-5,2.224-5,4.941V16H4.941C2.224,16,0,18.282,0,21
s2.224,5,4.941,5H16v11.059C16,39.776,18.282,42,21,42s5-2.224,5-4.941V26h11.059C39.776,26,42,23.718,42,21S39.776,16,37.059,16z"
        style={
          Object {
            "fill": "#fff",
          }
        }
      />
      <g />
      <g />
      <g />
      <g />
      <g />
      <g />
      <g />
      <g />
      <g />
      <g />
      <g />
      <g />
      <g />
      <g />
      <g />
    </svg>
  </button>
  <div
    className="journal__entries"
  />
</aside>
`;
